[BUG] Fix random seed consistency

a-hurst opened the pull request:
# PR Description

This PR is meant to fix the secondary issue discovered in issue #41: that is, the inability to get consistent robust re-referencing results with a random seed if different numbers of chunks were used across runs. I've tested it locally with data and it's consistent across different manually-set chunk sizes, so I *think* it's working okay, but please let me know if you can think of any problems with this approach!

# Merge Checklist

To merge your PR we need to first take the following points into account:

- [ ] the PR has been reviewed and all comments are resolved
- [ ] all [CI][what-is-ci] checks pass
- [ ] (if applicable): the PR description includes the phrase `closes #<issue-number>` to [automatically close an issue][auto-close-documentation]
- [ ] (if applicable): bug fixes, new features, or [API][what-is-api] changes are documented in [whats_new.rst][whats-new-file]

[contrib-guideline]: https://github.com/sappelhoff/pyprep/blob/master/.github/CONTRIBUTING.md

[what-is-ci]: https://help.github.com/en/actions/building-and-testing-code-with-continuous-integration/about-continuous-integration

[auto-close-documentation]: https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

[what-is-api]: https://en.wikipedia.org/wiki/Application_programming_interface

[whats-new-file]: https://github.com/sappelhoff/pyprep/blob/master/docs/whats_new.rst

a-hurst commented:
Huh, I guess one of the tests passes `n_samples` to `find_bads_by_ransac` as a float. I just changed the code to always coerce `n_samples` to an integer to make sure everything else checks out, but should I add proper type-checking instead? Or is this fix fine?

sappelhoff commented:
Did you use

```bash
git clone https://github.com/sappelhoff/pyprep
cd pyprep
pip install -r requirements-dev.txt
pre-commit install
pip install -e .
```

as described in https://github.com/sappelhoff/pyprep#installation ?

sappelhoff commented:
>  but should I add proper type-checking instead?

a proper check would be much nicer if it provides a reasonable error

sappelhoff commented on pyprep/find_noisy_channels.py:
```diff
@@ -586,19 +599,17 @@ def get_ransac_pred(self, chn_pos, chn_pos_good, good_chn_labs, n_pred_chns, dat
             channel numbers used for interpolation for RANSAC
         data : ndarray
             2-D EEG data
+        sample: int
```
in numpydoc style, there is a space before the `:`

```suggestion
        sample : int
```

sappelhoff commented on pyprep/find_noisy_channels.py:
```diff
@@ -68,6 +68,7 @@ def __init__(self, raw, do_detrend=True, random_state=None):
 
         # random_state
         self.random_state = check_random_state(random_state)
+        self.random_ch_picks = []
```
I'd add a comment here

```suggestion
        self.random_ch_picks = []  # needed for ransac
```

sappelhoff commented:
The general idea looks good to me :thinking: I'll let @yjmantilla chime in before merging.

@a-hurst thanks for your first PR to pyprep :tada: can you please also add a changelog entry and add yourself to the list of authors? (including an RST link to a page that you want users to find if they click on your name)

https://github.com/sappelhoff/pyprep/blob/master/docs/whats_new.rst

yjmantilla reviewed (commented):
The only thing that bugs me is that (I think) we can remove the ``n_pred_chns`` parameter in:

- self.ransac_correlations()
- self.self.run_ransac()
- self.get_ransac_pred()

I mean one passes it to the other untit get_ransac_pred just trashes it.

So I guess that for staying on the good programming practices we should remove it from the code and the functions documentations.

Now regarding the pytest stuff, Im looking at it to see what went wrong.

What do you think?

a-hurst commented:
Good catch @yjmantilla, just implemented the suggestion! As for the tests, I'm wondering if it's just a fluke or whether I've somehow broken, hopefully the new push will let us know what's what. If it continues, we can add the `-s` flag to pytest and get the full printed RANSAC output to help debug.

yjmantilla commented on pyprep/find_noisy_channels.py:
```diff
@@ -438,6 +444,15 @@ def find_bad_by_ransac(
+        rng = check_random_state(self.random_state)
+        for i in range(n_samples):
+            # Pick a random subset of clean channels to use for interpolation
+            good_chans = np.arange(chn_pos_good.shape[0])
```
good_chans could be out of the loop right? I mean Im sure the impact is minimal but you know...

yjmantilla reviewed (commented):
Just a minor suggestion...

sappelhoff commented on docs/whats_new.rst:
```diff
@@ -32,7 +33,7 @@ Current
 Changelog
 ~~~~~~~~~
-- Nothing yet
+- Fixed RANSAC to give consistent results with a fixed seed across different chunk sizes, by `Austin Hurst`_ and `Yorguin Mantilla`_ (`#43 <https://github.com/sappelhoff/pyprep/pull/43>`_)
```
This should go into the "bug" section

```suggestion
- Nothing yet

Bug
~~~
- Fixed RANSAC to give consistent results with a fixed seed across different chunk sizes, by `Austin Hurst`_ and `Yorguin Mantilla`_ (`#43 <https://github.com/sappelhoff/pyprep/pull/43>`_)
```

Also, we'll probably have to solve a git conflict with your two PRs editing "whatsnew" (not an issue, just a heads up)

sappelhoff reviewed (commented):
I think the test failures may be due to GitHub Actions capping the amount of RAM we are allowed to use :thinking: 

are the tests passing locally for you?

a-hurst commented:
lol so I figured out the CI issue: because we're now pre-generating random channel picks for each sample, and we're testing the memory error exception by asking it to do 10^100 samples, it gets stuck generating a list of billions upon billions of picks and the process eventually gets killed due to RAM consumption.

I've fixed the issue by adding another RAM check prior to the channel picks so that we don't generate random samples for impossibly-high sample counts. To avoid reduplicating code and clean things up a bit, I moved the RAM check stuff to its own separate function, also implementing the max ram proportion from the other PR while I'm at it. Hopefully everything should work now!

sappelhoff commented on docs/whats_new.rst:
```diff
@@ -34,6 +35,10 @@ Changelog
 
+Bug
+~~~
+- Fixed RANSAC to give consistent results with a fixed seed across different chunk sizes, by `Austin Hurst`_ and `Yorguin Mantilla`_ (`#43 <https://github.com/sappelhoff/pyprep/pull/43>`_)
```
actually, can you please add an additional (new) "API" section underneath the "bug" section where you mention that the parameters have changed for RANSAC methods? i.e., `n_pred_chns` has been dropped from methods X, Y, Z, ... It should refer to the same PR.

```RST
API
~~~
- did that and that .., by `Austin Hurst`_ and `Yorguin Mantilla`_ (`#43 <https://github.com/sappelhoff/pyprep/pull/43>`_)


```

sappelhoff reviewed (commented):
Haha, okay good that you caught it. Can you now rebase this PR on `master` and resolve the merge conflicts, please? 

I also added one more comment for the changelog. Other than that, I think this PR is good to be merged as well.

a-hurst commented:
Ah, whoops, messed the rebase process up. Let me try again.

a-hurst commented:
@sappelhoff Okay, everything should be good to go now! I also fixed some warnings I was getting from sphinxdoc when testing the changes I made to all the docstrings.

sappelhoff commented on docs/whats_new.rst:
```diff
@@ -35,6 +35,14 @@ Changelog
+
+API
+~~~
+- Added new function :func:`utils.verify_free_ram` and removed `n_samples` argument from the :meth:`NoisyChannels.ransac_correlations`, :meth:`NoisyChannels.run_ransac`, and :meth:`NoisyChannels.get_ransac_pred` methods, by `Austin Hurst`_ and `Yorguin Mantilla`_ (`#43 <https://github.com/sappelhoff/pyprep/pull/43>`_)
```
What we are writing in the API docs is meant for users, so if you write that there is a new function `verify_free_ram` you build some kind of expectation that users can/should use this, and that the API is relatively stable. And finally, that it's documented in the API section of the docs.

I suggest we keep `verify_free_ram` private for now, because users do not need to interact with it. Technically that means that we could/should rename the function to start with a leading underscore to follow python convention: `_verify_free_ram`, but it'd be okay with me to skip that for now.

However I recommend to remove that sentence from the user-facing changelog. The exact parameters that changed for the class methods are a bit too much I would say - I think it'd suffice to have a sentence like this:

```suggestion
- The permissible parameters for the following methods were removed and/or reordered: :meth:`NoisyChannels.ransac_correlations`, :meth:`NoisyChannels.run_ransac`, and :meth:`NoisyChannels.get_ransac_pred`, by `Austin Hurst`_ and `Yorguin Mantilla`_ (`#43 <https://github.com/sappelhoff/pyprep/pull/43>`_)
```

sappelhoff commented on docs/api.rst:
```diff
@@ -23,7 +23,7 @@ The :class:`NoisyChannels` class
    NoisyChannels.find_bad_by_nan_flat
    NoisyChannels.find_bad_by_deviation
-   NoisyChannels.find_bad_by_hf_noise
+   NoisyChannels.find_bad_by_hfnoise
```
good catch!

sappelhoff commented on pyprep/prep_pipeline.py:
```diff
@@ -21,16 +21,17 @@ class PrepPipeline:
-            - list of floats indicating frequencies to be removed.
-              For example for 60Hz you may specify ``np.arange(60, sfreq / 2, 60)``.
-              Specify an empty list to skip the line noise removal step.
+            - ref_chs : list | 'eeg'
+                - A list of channel names to be used for rereferencing.
+                  Can be a str 'eeg' to use all EEG channels.
+            - reref_chs : list | 'eeg'
+                - A list of channel names to be used for line-noise removed,
+                  and referenced. Can be a str 'eeg' to use all EEG channels.
+            - line_freqs : array_like
+                - list of floats indicating frequencies to be removed.
+                  For example, for 60Hz you may specify
+                  ``np.arange(60, sfreq / 2, 60)``. Specify an empty list to
+                  skip the line noise removal step.
     montage : DigMontage
```
Inspecting the [current docs](https://pyprep.readthedocs.io/en/latest/generated/prep_pipeline.PrepPipeline.html), it's clear that something is wrong:

![image](https://user-images.githubusercontent.com/9084751/101208136-fedc8000-3671-11eb-9677-60bce1ca3625.png)


However, your [changes](https://pyprep--43.org.readthedocs.build/en/43/generated/prep_pipeline.PrepPipeline.html#prep_pipeline.PrepPipeline) seem to make a part of the docstr bold that should not be bold:

![image](https://user-images.githubusercontent.com/9084751/101208252-292e3d80-3672-11eb-86c0-d41c31fc1722.png)


--- 

I am not sure, but would it have been sufficient to just add a blank line right after line 23, instead of increasing the indentation level?

a-hurst replied:
Yep, that did it. Thanks!

sappelhoff commented on pyprep/utils.py:
```diff
@@ -150,3 +151,39 @@ def make_random_mne_object(
+        Number of samples to use for computation of ransac.
+    n_channels : int
+        Number of channels to process per chunk.
+    max_prop: float
```
```suggestion
    max_prop : float
```

I wonder why [flake8-docstrings](https://pypi.org/project/flake8-docstrings/) does not catch this

a-hurst replied:
Yeah, weird... I keep forgetting the spaces on these because I've only ever used Google-style Python docstrings before, gotta fight some well-learned reflexes to conform to numpydoc!

sappelhoff reviewed (approved):
Glad the rebase worked out at the second try :) 

some minor nitpicks, then we are good to go!

a-hurst commented:
@sappelhoff Okay, hopefully that does it!

sappelhoff commented:
Thanks a lot @a-hurst!
